budgetkey_data_pipelines/pipelines/budget/national/changes/processed/extract-change-groups.py

```diff
@@ -0,0 +1,237 @@
+                    try:
+                        while True:
+                            i += 1
+                            if i % 100000 == 0:
+                                logging.debug('%s %s %s' % (
+                                    len(date_reserve), num_found, i
+                                ))
+                            transaction = date_transactions.send(found)
+                            found = False
+                            sumvec = sum(
+                                sum(c.get(x, 0) for c in transaction) ** 2
+                                for x in value_fields
+                            )
+                            if sumvec == 0:
+                                transfer_codes = set(
+                                    x['trcode']
+                                    for x in transaction
+                                )
+                                if len(selected_transfer_codes & transfer_codes) > 0:
+                                    continue
+                                selected_transfer_codes.update(transfer_codes)
+                                transfer_codes = sorted(transfer_codes)
+                                num_found += len(transfer_codes)
+                                yield set(transfer_codes),
+                                found = True
+                    except StopIteration:
+                        pass
+        for change in changes:
+            if not change['trcode'] in selected_transfer_codes:
+                selected_transfer_codes.add(change['trcode'])
+                yield {change['trcode']}
+
+    def assign_transactions(groups, changes):
+
+        # This allows to perform search ~10 times faster
+        changes_by_year = {}
+        for x in changes:
+            year = x['year']
+            changes_by_year[year] = changes_by_year.get(year, [])
+            changes_by_year[year].append(x)
+
+        processed_count = 0
+        for trcodes in groups:
+            # Just a check
+            years = set(
+                int(x.split('/')[0])
+                for x in trcodes
+            )
+            assert(len(years) == 1)
+            year = years.pop()
+
+            transfer_changes = list(filter(
+                lambda x: x['trcode'] in trcodes,
+                changes_by_year[year]
+            ))
+
+            budget_codes = set(
+                x['budget_code']
+                for x in transfer_changes
+            )
+            prefixes = sorted(set(
+                itertools.chain.from_iterable(
+                    [code[:l] for l in range(2, 10, 2)]
+                    for code in budget_codes
+                )
+            ), key=lambda x: int('1' + x))
+
+            transfer_ids = list(set(
+                x.split('/')[1]
+                for x in trcodes
+            ))
+
+            transaction = {
+                'id': transfer_ids[0],
+                'changes': get_keys_for_prefixes(transfer_changes, prefixes),
```

Comment:
See the issue - we just need 3 columns:
- leading_item
- req_code
- transaction_id
with one row per transaction_id.

So I think this code can be greatly simplified.